instrumentation/servlet/servlet-3.0/javaagent/src/test/groovy/AbstractServlet3Test.groovy

```diff
@@ -45,6 +47,25 @@ abstract class AbstractServlet3Test<SERVER, CONTEXT> extends HttpServerTest<SERV
     addServlet(context, INDEXED_CHILD.path, servlet)
   }
 
+  def cleanup() {
+    // wait for async request threads to complete
+    await()
+      .atMost(15, TimeUnit.SECONDS)
+      .until({ !isRequestRunning() })
+  }
+
+  static boolean isRequestRunning() {
+    def result = Thread.getAllStackTraces().values().find {stackTrace ->
+      def element = stackTrace.find {
+        return ((it.className == "org.apache.catalina.core.AsyncContextImpl\$RunnableWrapper" && it.methodName == "run")
```

Comment:
I'm a little scared of adding too many of these sort of checks. Have we ruled out

- Test, client call doesn't actually consume entire response or otherwise ends test before request completes properly

- Instrumentation is not ordering context close, span finish, response return, correctly

Reply:
My understanding is that async request is completed when response is sent to client as this happens inside the async task it hasn't exited it's scope yet. Usually async task will manage to exit before test completes asserting spans and does context leak check but as our tests generate high load it can happen that everything else completes before async context runnable exits.
Build with failures on tomcat https://github.com/open-telemetry/opentelemetry-java-instrumentation/runs/3407877673
I hope we can find some other way of solving this as the same issue will affect a lot of tests.